### PR TITLE
[FR] Support Rule Alert Suppression in Rule Schema

### PR DESCRIPTION
**Pull request by terrancedejesus**
## Issues
* https://github.com/elastic/detection-rules/issues/2657

## Summary
Adds rule alert suppression support in rule schema. Please see issue for more details.

## Additional Information

API format should resemble the following. If this is not a "Per Time" suppression then `duration` does not simply exist.

```sql
  "alert_suppression":{
    "group_by":[
      "process.parent.code_signature.trusted"
    ],
    "duration":{
      "unit":"m",
      "value":10
    }
```

<details>
<summary>Example Rule with Per Time</summary>

```sql
[metadata]
creation_date = "2022/05/16"
integration = ["endpoint"]
maturity = "production"
min_stack_comments = "New fields added: required_fields, related_integrations, setup"
min_stack_version = "8.3.0"
updated_date = "2023/02/22"

[rule]
author = ["Elastic"]
description = """
Identifies an outbound network connection attempt followed by a session id change as the root user by the same process
entity. This particular instantiation of a network connection is abnormal and should be investigated as it may indicate
a potential reverse shell activity via a privileged process.
"""
false_positives = [
    """
    False-Positives (FP) can appear if another remote terminal service is being used to connect to it's listener but
    typically SSH is used in these scenarios.
    """,
]
from = "now-9m"
index = ["logs-endpoint.events.*"]
language = "eql"
license = "Elastic License v2"
name = "Suspicious Network Connection Attempt by Root"
note = """## Triage and analysis
### Investigating Connection Attempt by Non-SSH Root Session
Detection alerts from this rule indicate a strange or abnormal outbound connection attempt by a privileged process.  Here are some possible avenues of investigation:
- Examine unusual and active sessions using commands such as 'last -a', 'netstat -a', and 'w -a'.
- Analyze processes and command line arguments to detect anomalous process execution that may be acting as a listener.
- Analyze anomalies in the use of files that do not normally initiate connections.
- Examine processes utilizing the network that do not normally have network communication.
"""
references = [
    "https://www.sandflysecurity.com/blog/linux-file-masquerading-and-malicious-pids-sandfly-1-2-6-update/",
    "https://twitter.com/GossiTheDog/status/1522964028284411907",
    "https://exatrack.com/public/Tricephalic_Hellkeeper.pdf",
]
risk_score = 47
rule_id = "eb6a3790-d52d-11ec-8ce9-f661ea17fbce"
severity = "medium"
tags = ["Elastic", "Host", "Linux", "Threat Detection", "Command and Control"]
type = "eql"

query = '''
sequence by process.entity_id with maxspan=1m
[network where host.os.type == "linux" and event.type == "start" and event.action == "connection_attempted" and user.id == "0" and
    not process.executable : ("/bin/ssh", "/sbin/ssh", "/usr/lib/systemd/systemd", "/usr/sbin/sshd")]
[process where host.os.type == "linux" and event.action == "session_id_change" and user.id == "0" and
    not process.executable : ("/bin/ssh", "/sbin/ssh", "/usr/lib/systemd/systemd", "/usr/sbin/sshd")]
'''


[[rule.threat]]
framework = "MITRE ATT&CK"
[[rule.threat.technique]]
id = "T1095"
name = "Non-Application Layer Protocol"
reference = "https://attack.mitre.org/techniques/T1095/"


[rule.threat.tactic]
id = "TA0011"
name = "Command and Control"
reference = "https://attack.mitre.org/tactics/TA0011/"
[[rule.threat]]
framework = "MITRE ATT&CK"
[[rule.threat.technique]]
id = "T1548"
name = "Abuse Elevation Control Mechanism"
reference = "https://attack.mitre.org/techniques/T1548/"
[[rule.threat.technique.subtechnique]]
id = "T1548.003"
name = "Sudo and Sudo Caching"
reference = "https://attack.mitre.org/techniques/T1548/003/"



[rule.threat.tactic]
id = "TA0004"
name = "Privilege Escalation"
reference = "https://attack.mitre.org/tactics/TA0004/"

[rule.alert_suppression]
group_by = ["process.name"]

[rule.alert_suppression.duration]
unit = "m"
value = 10

```

</details>

<details>
<summary>Example API Format for Per Time</summary>

```sql
{
  "alert_suppression": {
    "duration": {
      "unit": "m",
      "value": 10
    },
    "group_by": [
      "process.parent.code_signature.trusted"
    ]
  },
  "author": [
    "Elastic"
  ],
  "description": "Identifies an outbound network connection attempt followed by a session id change as the root user by the same process entity. This particular instantiation of a network connection is abnormal and should be investigated as it may indicate a potential reverse shell activity via a privileged process.",
  "false_positives": [
    "False-Positives (FP) can appear if another remote terminal service is being used to connect to it's listener but typically SSH is used in these scenarios."
  ],
  "from": "now-9m",
  "index": [
    "logs-endpoint.events.*"
  ],
  "language": "eql",
  "license": "Elastic License v2",
  "name": "Suspicious Network Connection Attempt by Root",
  "note": "## Triage and analysis\n### Investigating Connection Attempt by Non-SSH Root Session\nDetection alerts from this rule indicate a strange or abnormal outbound connection attempt by a privileged process.  Here are some possible avenues of investigation:\n- Examine unusual and active sessions using commands such as 'last -a', 'netstat -a', and 'w -a'.\n- Analyze processes and command line arguments to detect anomalous process execution that may be acting as a listener.\n- Analyze anomalies in the use of files that do not normally initiate connections.\n- Examine processes utilizing the network that do not normally have network communication.\n",
  "query": "sequence by process.entity_id with maxspan=1m\n[network where host.os.type == \"linux\" and event.type == \"start\" and event.action == \"connection_attempted\" and user.id == \"0\" and\n    not process.executable : (\"/bin/ssh\", \"/sbin/ssh\", \"/usr/lib/systemd/systemd\", \"/usr/sbin/sshd\")]\n[process where host.os.type == \"linux\" and event.action == \"session_id_change\" and user.id == \"0\" and\n    not process.executable : (\"/bin/ssh\", \"/sbin/ssh\", \"/usr/lib/systemd/systemd\", \"/usr/sbin/sshd\")]\n",
  "references": [
    "https://www.sandflysecurity.com/blog/linux-file-masquerading-and-malicious-pids-sandfly-1-2-6-update/",
    "https://twitter.com/GossiTheDog/status/1522964028284411907",
    "https://exatrack.com/public/Tricephalic_Hellkeeper.pdf"
  ],
  "related_integrations": [
    {
      "package": "endpoint",
      "version": "^8.2.0"
    }
  ],
  "required_fields": [
    {
      "ecs": true,
      "name": "event.action",
      "type": "keyword"
    },
    {
      "ecs": true,
      "name": "event.type",
      "type": "keyword"
    },
    {
      "ecs": true,
      "name": "host.os.type",
      "type": "keyword"
    },
    {
      "ecs": true,
      "name": "process.entity_id",
      "type": "keyword"
    },
    {
      "ecs": true,
      "name": "process.executable",
      "type": "keyword"
    },
    {
      "ecs": true,
      "name": "user.id",
      "type": "keyword"
    }
  ],
  "risk_score": 47,
  "rule_id": "eb6a3790-d52d-11ec-8ce9-f661ea17fbce",
  "severity": "medium",
  "tags": [
    "Elastic",
    "Host",
    "Linux",
    "Threat Detection",
    "Command and Control"
  ],
  "threat": [
    {
      "framework": "MITRE ATT&CK",
      "tactic": {
        "id": "TA0011",
        "name": "Command and Control",
        "reference": "https://attack.mitre.org/tactics/TA0011/"
      },
      "technique": [
        {
          "id": "T1095",
          "name": "Non-Application Layer Protocol",
          "reference": "https://attack.mitre.org/techniques/T1095/"
        }
      ]
    },
    {
      "framework": "MITRE ATT&CK",
      "tactic": {
        "id": "TA0004",
        "name": "Privilege Escalation",
        "reference": "https://attack.mitre.org/tactics/TA0004/"
      },
      "technique": [
        {
          "id": "T1548",
          "name": "Abuse Elevation Control Mechanism",
          "reference": "https://attack.mitre.org/techniques/T1548/",
          "subtechnique": [
            {
              "id": "T1548.003",
              "name": "Sudo and Sudo Caching",
              "reference": "https://attack.mitre.org/techniques/T1548/003/"
            }
          ]
        }
      ]
    }
  ],
  "type": "eql",
  "version": 102
}
```

</details>

## Testing
To test do the following steps:
1. Pull updates and checkout this branch
2. Add the example rule locally
3. Run `python -m detection_rules view-rule /Users/tdejesus/code/src/detection-rules/rules/linux/RULEPATH`

Expected results should be the API format of the rule in the console. To test the validation, you can key:value pairs that are not explicitly allowed.

### CODE REVIEW

**Comment by terrancedejesus** [comment]
@brokensound77 @Mikaayenson @eric-forte-elastic - I can abstract the duration code a bit more in the schema so it `duration` is nested inside of `alert_suppression` like `rule.alert_suppression.duration`. Just let me know.

**Comment by Mikaayenson** [comment]
@terrancedejesus 
- Will validate the check to group by max 3 Screenshot 2023-03-23 at 4 12 55 PM
- We probably should validate the fields as well to make sure they are ecs / valid fields

**Comment by terrancedejesus** [comment]
> We probably should validate the fields as well to make sure they are ecs / valid fields
Yes, I plan to add that with the unit tests. Good point.

**Comment by brokensound77** [comment]
> > We probably should validate the fields as well to make sure they are ecs / valid fields
> 
> Yes, I plan to add that with the unit tests. Good point.

++ To unit tests rather than within validation

**Comment by brokensound77** [comment]
> 

The problem with limiting it only to ECS fields is when a rule has non-ecs fields, like most of the integrations. Rather than full validation in unit tests, should be able to get away fairly cheaply by doing something like:

```python
group_fields = set(group_by)
combined_fields = set([ecs] + [integration])

# validate
group_fields & combined_fields == group_fields
```

File: detection_rules/rule.py [hunk 1]
```diff
@@ -177,6 +189,7 @@ class RelatedIntegrations:
         integration: Optional[definitions.NonEmptyStr]
 
     actions: Optional[list]
+    alert_suppression: Optional[AlertSuppressionMapping]
```
[the remark this answers is not in the document]
Comment (terrancedejesus) [reply detection_rules/rule.py:192]: Your right it was 8.6 but "per time" was introduced in 8.7.

**Review by eric-forte-elastic — APPROVED** [review]
LGTM :+1:

**Review by Mikaayenson — APPROVED** [review]
LGTM. Based on internal discussion, we decided the structure would be nested to be more consistent with other fields.

**Comment by terrancedejesus** [comment]
##### Update 03/27/2023
During our final review, we attempted to import the rule into an existing 8.7.0 snapshot stack and received a "invalid keys" error from the UI. Through testing and UI exploration, we learned that alert suppression is only available for KQL rules where we were testing with an EQL rule and ran into errors. Since alert suppression is not a rule type, KQL verification has to come from unit tests where we verify the value of `language` in the rule contents. Below are a couple screenshots.

Unit test failing because language is EQL.
Screenshot 2023-03-27 at 3 09 40 PM

Unit test passing with the same rule because it is valid KQL.
Screenshot 2023-03-27 at 3 12 04 PM

Successful import of rule after changed to KQL.
Screenshot 2023-03-27 at 3 15 16 PM

**Pull request merged**